File: src/git_adr/core/templates.py
"""Template Engine for ADR formats.

This module provides template rendering and format conversion for
various ADR formats:
- MADR 4.0 (default) - Full template with options analysis
- Nygard - Original minimal format
- Y-Statement - Single-sentence decision format
- Alexandrian - Pattern-language inspired
- Business Case - Extended business justification
- Planguage - Quality-focused measurable format
- Custom - User-provided templates
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from git_adr.core.adr import ADR

# =============================================================================
# Template Definitions
# =============================================================================

MADR_TEMPLATE = """# {title}

## Status

{status}

## Context

<!-- Describe the context and problem statement -->

## Decision

<!-- Describe the decision that was made -->

## Consequences

### Positive

<!-- List positive consequences -->

### Negative

<!-- List negative consequences -->

### Neutral

<!-- List neutral consequences -->

## Options Considered

### Option 1: [Name]

<!-- Description, pros, cons -->

### Option 2: [Name]

<!-- Description, pros, cons -->

## Decision Outcome

Chosen option: "[option name]", because [justification].

## More Information

<!-- Links, references, related ADRs -->
"""

NYGARD_TEMPLATE = """# {title}

## Status

{status}

## Context

<!-- What is the issue that we're seeing that is motivating this decision or change? -->

## Decision

<!-- What is the change that we're proposing and/or doing? -->

## Consequences

<!-- What becomes easier or more difficult to do because of this change? -->
"""

Y_STATEMENT_TEMPLATE = """# {title}

## Status

{status}

## Decision

In the context of [context],
facing [concern],
we decided [decision],
to achieve [goal],
accepting [tradeoff].

## More Information

<!-- Additional context or references -->
"""

ALEXANDRIAN_TEMPLATE = """# {title}

## Status

{status}

## Context

<!-- The situation that gives rise to the need for this decision -->

## Forces

<!-- The competing constraints and concerns -->

## Problem

<!-- The specific problem or question being addressed -->

## Solution

<!-- The decision or approach chosen -->

## Resulting Context

<!-- The situation after applying the solution -->

## Related Patterns

<!-- Links to related ADRs or patterns -->
"""

BUSINESS_CASE_TEMPLATE = """# {title}

## Status

{status}

## Executive Summary

<!-- Brief overview for stakeholders -->

## Business Context

<!-- Business drivers and strategic alignment -->

## Problem Statement

<!-- Clear description of the problem -->

## Proposed Solution

<!-- The recommended approach -->

## Options Analysis

### Option 1: [Name]

**Cost:** [estimate]
**Time:** [estimate]
**Risk:** [Low/Medium/High]

**Pros:**
-

**Cons:**
-

### Option 2: [Name]

**Cost:** [estimate]
**Time:** [estimate]
**Risk:** [Low/Medium/High]

**Pros:**
-

**Cons:**
-

## Financial Impact

<!-- Cost/benefit analysis, ROI -->

## Risk Assessment

<!-- Identified risks and mitigation strategies -->

## Implementation Plan

<!-- High-level implementation approach -->

## Approval

| Role | Name | Date | Decision |
|------|------|------|----------|
| Sponsor | | | |
| Architect | | | |
| Tech Lead | | | |
"""

PLANGUAGE_TEMPLATE = """# {title}

## Status

{status}

## Tag

{id}

## Gist

<!-- One-sentence summary of the decision -->

## Background

<!-- Context and motivation -->

## Scale

<!-- Measurement scale for success -->

## Meter

<!-- How success will be measured -->

## Past

<!-- Current state/baseline -->

## Must

<!-- Minimum acceptable level (hard constraint) -->

## Plan

<!-- Target level (expected outcome) -->

## Wish

<!-- Ideal level (stretch goal) -->

## Assumptions

<!-- Key assumptions underlying this decision -->

## Risks

<!-- Potential risks and mitigation -->
"""


# =============================================================================
# Template Registry
# =============================================================================

TEMPLATES: dict[str, str] = {
    "madr": MADR_TEMPLATE,
    "nygard": NYGARD_TEMPLATE,
    "y-statement": Y_STATEMENT_TEMPLATE,
    "alexandrian": ALEXANDRIAN_TEMPLATE,
    "business": BUSINESS_CASE_TEMPLATE,
    "planguage": PLANGUAGE_TEMPLATE,
}

TEMPLATE_DESCRIPTIONS: dict[str, str] = {
    "madr": "MADR 4.0 - Full template with options analysis (default)",
    "nygard": "Nygard - Original minimal ADR format",
    "y-statement": "Y-Statement - Single-sentence decision format",
    "alexandrian": "Alexandrian - Pattern-language inspired format",
    "business": "Business Case - Extended business justification",
    "planguage": "Planguage - Quality-focused measurable format",
}


# =============================================================================
# Template Engine
# =============================================================================


@dataclass
class TemplateContext:
    """Context for template rendering.

    Contains all the variables available for substitution in templates.
    """

    id: str
    title: str
    date: date
    status: str
    tags: list[str]
    deciders: list[str]
    supersedes: str | None = None

    def to_dict(self) -> dict[str, str]:
        """Convert to dictionary for template substitution.

        Returns:
            Dictionary with string values.
        """
        return {
            "id": self.id,
            "title": self.title,
            "date": self.date.isoformat(),
            "status": self.status,
            "tags": ", ".join(self.tags) if self.tags else "",
            "deciders": ", ".join(self.deciders) if self.deciders else "",
            "supersedes": self.supersedes or "",
        }


class TemplateEngine:
    """Engine for rendering and converting ADR templates.

    Supports built-in formats and custom templates loaded from files.
    """

    def __init__(self, custom_templates_dir: Path | None = None) -> None:
        """Initialize the template engine.

        Args:
            custom_templates_dir: Directory containing custom templates.
        """
        self._templates = dict(TEMPLATES)
        self._custom_dir = custom_templates_dir

        # Load custom templates if directory exists
        if custom_templates_dir and custom_templates_dir.exists():
            self._load_custom_templates(custom_templates_dir)

    def _load_custom_templates(self, directory: Path) -> None:
        """Load custom templates from a directory.

        Custom templates should be named {format_name}.md

        Args:
            directory: Directory containing template files.
        """
        for template_file in directory.glob("*.md"):
            format_name = template_file.stem
            try:
                template_content = template_file.read_text()
                self._templates[format_name] = template_content
            except OSError:
                continue

    def list_formats(self) -> list[str]:
        """List all available format names.

        Returns:
            List of format names.
        """
        return sorted(self._templates.keys())

    def get_template(self, format_name: str) -> str | None:
        """Get a template by format name.

        Args:
            format_name: Name of the format.

        Returns:
            Template string, or None if not found.
        """
        return self._templates.get(format_name)

    def render(
        self,
        format_name: str,
        context: TemplateContext,
    ) -> str:
        """Render a template with the given context.

        Args:
            format_name: Name of the format to render.
            context: Template context with values.

        Returns:
            Rendered template content.

        Raises:
            ValueError: If format is not found.
        """
        template = self.get_template(format_name)
        if template is None:
            available = ", ".join(self.list_formats())
            raise ValueError(f"Unknown format: {format_name}. Available: {available}")

        # Substitute context values
        content = template.format(**context.to_dict())

        return content

    def render_for_new(
        self,
        format_name: str,
        title: str,
        adr_id: str,
        status: str = "proposed",
        tags: list[str] | None = None,
        deciders: list[str] | None = None,
        supersedes: str | None = None,
    ) -> str:
        """Render a template for a new ADR.

        Convenience method that creates the context and renders.

        Args:
            format_name: Format name.
            title: ADR title.
            adr_id: Generated ADR ID.
            status: Initial status.
            tags: Optional tags.
            deciders: Optional deciders.
            supersedes: ID of superseded ADR.

        Returns:
            Rendered template content.
        """
        context = TemplateContext(
            id=adr_id,
            title=title,
            date=date.today(),
            status=status,
            tags=tags or [],
            deciders=deciders or [],
            supersedes=supersedes,
        )

        return self.render(format_name, context)

    def detect_format(self, content: str) -> str:
        """Detect the format of an existing ADR.

        Uses heuristics to identify the format based on content structure.

        Args:
            content: ADR content (without frontmatter).

        Returns:
            Detected format name, or "unknown".
        """
        content_lower = content.lower()

        # Check for distinctive sections
        if "## options considered" in content_lower:
            return "madr"
        if "## decision outcome" in content_lower:
            return "madr"

        if "in the context of" in content_lower and "we decided" in content_lower:
            return "y-statement"

        if "## forces" in content_lower and "## resulting context" in content_lower:
            return "alexandrian"

        if "## financial impact" in content_lower or "## approval" in content_lower:
            return "business"

        if "## scale" in content_lower and "## meter" in content_lower:
            return "planguage"

        # Default to nygard if has basic structure
        if "## context" in content_lower and "## decision" in content_lower:
            return "nygard"

        return "unknown"

    def convert(
        self,
        adr: ADR,
        target_format: str,
    ) -> str:
        """Convert an ADR to a different format.

        Attempts to preserve content while restructuring to the target format.

        Args:
            adr: Source ADR.
            target_format: Target format name.

        Returns:
            Converted content.

        Raises:
            ValueError: If target format is not found.
        """
        template = self.get_template(target_format)
        if template is None:
            raise ValueError(f"Unknown format: {target_format}")

        # Extract sections from source content
        sections = self._extract_sections(adr.content)

        # Create context from metadata
        context = TemplateContext(
            id=adr.metadata.id,
            title=adr.metadata.title,
            date=adr.metadata.date,
            status=str(adr.metadata.status),
            tags=adr.metadata.tags,
            deciders=adr.metadata.deciders,
            supersedes=adr.metadata.supersedes,
        )

        # Render base template
        content = self.render(target_format, context)

        # Merge extracted sections into the new format
        content = self._merge_sections(content, sections, target_format)

        return content

    def _extract_sections(self, content: str) -> dict[str, str]:
        """Extract sections from ADR content.

        Args:
            content: ADR markdown content.

        Returns:
            Dictionary mapping section name to content.
        """
        import re

        sections: dict[str, str] = {}
        current_section = ""
        current_content: list[str] = []

        for line in content.split("\n"):
            # Check for section header
            header_match = re.match(r"^##\s+(.+)$", line)
            if header_match:
                # Save previous section
                if current_section:
                    sections[current_section.lower()] = "\n".join(
                        current_content
                    ).strip()

                current_section = header_match.group(1)
                current_content = []
            else:
                current_content.append(line)

        # Save last section
        if current_section:
            sections[current_section.lower()] = "\n".join(current_content).strip()

        return sections

    def _merge_sections(
        self,
        template_content: str,
        sections: dict[str, str],
        target_format: str,
    ) -> str:
        """Merge extracted sections into a template.

        Args:
            template_content: Rendered template.
            sections: Extracted sections from source.
            target_format: Target format name.

        Returns:
            Content with sections merged.
        """
        import re

        # Define section mappings for common sections
        # Maps source section names to target section names
        mappings: dict[str, list[str]] = {
            "context": ["context", "background", "business context"],
            "decision": ["decision", "proposed solution", "solution"],
            "consequences": ["consequences", "resulting context"],
            "problem": ["problem", "problem statement"],
            "options": ["options considered", "options analysis"],
        }

        result = template_content

        for source_name, source_content in sections.items():
            if not source_content or "<!--" in source_content:
                continue

            # Find matching target section
            for canonical, targets in mappings.items():
                if source_name in targets or canonical in source_name:
                    # Try to replace placeholder content in target
                    for target_name in targets:
                        pattern = (
                            rf"(## {re.escape(target_name.title())}.*?\n)(<!-- .+? -->)"
                        )
                        replacement = rf"\1{source_content}"
                        result, count = re.subn(
                            pattern,
                            replacement,
                            result,
                            flags=re.IGNORECASE | re.DOTALL,
                        )
                        if count > 0:
                            break

        return result


def get_default_template_engine() -> TemplateEngine:
    """Get the default template engine.

    Returns:
        TemplateEngine with built-in templates.
    """
    return TemplateEngine()


def render_initial_adr(
    adr_id: str = "00000000-use-adrs",
    title: str = "Use Architecture Decision Records",
) -> str:
    """Render the initial ADR created during init.

    This ADR documents the decision to use ADRs for the project.

    Args:
        adr_id: ADR ID.
        title: ADR title.

    Returns:
        Rendered ADR content.
    """
    return f"""# {title}

## Status

accepted

## Context

We need to record the architectural decisions made on this project.

## Decision

We will use Architecture Decision Records (ADRs) stored in git notes,
as described by git-adr. This approach:

- Keeps ADRs invisible in the working tree but visible in history
- Eliminates merge conflicts from sequential numbering
- Associates decisions with implementing commits
- Syncs automatically with regular git operations

## Consequences

### Positive

- Decisions are documented and searchable
- Context is preserved for future team members
- ADRs don't clutter the repository file tree
- Natural integration with git workflow

### Negative

- Team members need to learn git-adr commands
- Git notes require configuration for remote sync

### Neutral

- Existing file-based ADRs can be imported
"""

File: src/core/templates.rs
//! Template engine for ADR generation.
//!
//! This module provides template rendering using Tera,
//! with built-in templates for common ADR formats.

use crate::Error;
use std::collections::HashMap;
use tera::{Context, Tera};

/// Built-in ADR template: Nygard format.
pub const TEMPLATE_NYGARD: &str = r#"# {{ title }}

## Status

{{ status }}

## Context

{{ context | default(value="What is the issue that we're seeing that is motivating this decision or change?") }}

## Decision

{{ decision | default(value="What is the change that we're proposing and/or doing?") }}

## Consequences

{{ consequences | default(value="What becomes easier or more difficult to do because of this change?") }}
"#;

/// Built-in ADR template: MADR format.
pub const TEMPLATE_MADR: &str = r#"# {{ title }}

## Status

{{ status }}

{% if deciders %}
Deciders: {{ deciders | join(sep=", ") }}
{% endif %}
{% if date %}
Date: {{ date }}
{% endif %}

## Context and Problem Statement

{{ context | default(value="Describe the context and problem statement...") }}

## Decision Drivers

{% for driver in decision_drivers | default(value=[]) %}
* {{ driver }}
{% else %}
* Driver 1
* Driver 2
{% endfor %}

## Considered Options

{% for option in options | default(value=[]) %}
* {{ option }}
{% else %}
* Option 1
* Option 2
{% endfor %}

## Decision Outcome

Chosen option: "{{ chosen_option | default(value="Option X") }}"

### Consequences

#### Good

{% for item in good_consequences | default(value=[]) %}
* {{ item }}
{% else %}
* Good consequence 1
{% endfor %}

#### Bad

{% for item in bad_consequences | default(value=[]) %}
* {{ item }}
{% else %}
* Bad consequence 1
{% endfor %}

## More Information

{{ more_info | default(value="Additional information, links, references...") }}
"#;

/// Built-in ADR template: Y-statement format.
pub const TEMPLATE_Y_STATEMENT: &str = r#"# {{ title }}

## Status

{{ status }}

## Decision

In the context of {{ context | default(value="<use case/user story>") }},
facing {{ facing | default(value="<concern>") }},
we decided for {{ decision | default(value="<option>") }}
and against {{ against | default(value="<other options>") }},
to achieve {{ achieve | default(value="<quality>") }},
accepting {{ accepting | default(value="<downside>") }}.
"#;

/// Built-in ADR template: Alexandrian format.
pub const TEMPLATE_ALEXANDRIAN: &str = r#"# {{ title }}

## Status

{{ status }}

## Prologue

{{ prologue | default(value="Summary and context for this ADR") }}

## Problem Statement

{{ problem | default(value="The problem or question being addressed") }}

## Forces

{% for force in forces | default(value=[]) %}
* {{ force }}
{% else %}
* Force 1: Description
* Force 2: Description
{% endfor %}

## Solution

{{ solution | default(value="The chosen solution") }}

## Consequences

{{ consequences | default(value="Resulting context after applying the solution") }}

## Related Patterns

{% for pattern in related | default(value=[]) %}
* {{ pattern }}
{% endfor %}
"#;

/// Built-in ADR template: Business Case format.
pub const TEMPLATE_BUSINESS_CASE: &str = r#"# {{ title }}

## Status

{{ status }}

## Executive Summary

{{ executive_summary | default(value="Brief overview of the decision and its impact") }}

## Background

{{ background | default(value="Context and history leading to this decision") }}

## Problem Statement

{{ problem | default(value="The business problem being addressed") }}

## Proposed Solution

{{ solution | default(value="Recommended approach") }}

## Alternatives Considered

{% for alt in alternatives | default(value=[]) %}
### {{ alt.name }}
{{ alt.description }}
{% else %}
### Alternative 1
Description of alternative approach
{% endfor %}

## Cost-Benefit Analysis

### Costs
{% for cost in costs | default(value=[]) %}
* {{ cost }}
{% else %}
* Implementation cost
* Maintenance cost
{% endfor %}

### Benefits
{% for benefit in benefits | default(value=[]) %}
* {{ benefit }}
{% else %}
* Benefit 1
* Benefit 2
{% endfor %}

## Risk Assessment

{% for risk in risks | default(value=[]) %}
* {{ risk }}
{% else %}
* Risk 1: Mitigation strategy
{% endfor %}

## Implementation Plan

{{ implementation | default(value="High-level implementation approach") }}

## Success Metrics

{% for metric in metrics | default(value=[]) %}
* {{ metric }}
{% else %}
* Metric 1
{% endfor %}
"#;

/// Template engine for ADR generation.
#[derive(Debug)]
pub struct TemplateEngine {
    tera: Tera,
}

impl Default for TemplateEngine {
    fn default() -> Self {
        Self::new()
    }
}

impl TemplateEngine {
    /// Create a new template engine with built-in templates.
    #[must_use]
    pub fn new() -> Self {
        let mut tera = Tera::default();

        // Add built-in templates
        let _ = tera.add_raw_template("nygard", TEMPLATE_NYGARD);
        let _ = tera.add_raw_template("madr", TEMPLATE_MADR);
        let _ = tera.add_raw_template("y-statement", TEMPLATE_Y_STATEMENT);
        let _ = tera.add_raw_template("alexandrian", TEMPLATE_ALEXANDRIAN);
        let _ = tera.add_raw_template("business-case", TEMPLATE_BUSINESS_CASE);

        Self { tera }
    }

    /// Add a custom template.
    ///
    /// # Errors
    ///
    /// Returns an error if the template is invalid.
    pub fn add_template(&mut self, name: &str, content: &str) -> Result<(), Error> {
        self.tera
            .add_raw_template(name, content)
            .map_err(|e| Error::TemplateError {
                message: format!("Failed to add template '{name}': {e}"),
            })
    }

    /// Render a template with the given context.
    ///
    /// # Errors
    ///
    /// Returns an error if rendering fails.
    pub fn render(
        &self,
        template: &str,
        context: &HashMap<String, String>,
    ) -> Result<String, Error> {
        let mut tera_context = Context::new();
        for (key, value) in context {
            tera_context.insert(key, value);
        }

        self.tera
            .render(template, &tera_context)
            .map_err(|e| Error::TemplateError {
                message: format!("Failed to render template '{template}': {e}"),
            })
    }

    /// List available templates.
    #[must_use]
    pub fn list_templates(&self) -> Vec<String> {
        self.tera.get_template_names().map(String::from).collect()
    }

    /// Check if a template exists.
    #[must_use]
    pub fn has_template(&self, name: &str) -> bool {
        self.tera.get_template_names().any(|n| n == name)
    }

    /// Get template content.
    ///
    /// # Errors
    ///
    /// Returns an error if the template doesn't exist.
    pub fn get_template(&self, name: &str) -> Result<String, Error> {
        // Built-in templates
        match name {
            "nygard" => Ok(TEMPLATE_NYGARD.to_string()),
            "madr" => Ok(TEMPLATE_MADR.to_string()),
            "y-statement" => Ok(TEMPLATE_Y_STATEMENT.to_string()),
            "alexandrian" => Ok(TEMPLATE_ALEXANDRIAN.to_string()),
            "business-case" => Ok(TEMPLATE_BUSINESS_CASE.to_string()),
            _ => Err(Error::TemplateNotFound {
                name: name.to_string(),
            }),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_template_engine_new() {
        let engine = TemplateEngine::new();
        assert!(engine.has_template("nygard"));
        assert!(engine.has_template("madr"));
        assert!(!engine.has_template("nonexistent"));
    }

    #[test]
    fn test_template_engine_default() {
        let engine = TemplateEngine::default();
        assert!(engine.has_template("nygard"));
    }

    #[test]
    fn test_render_nygard() {
        let engine = TemplateEngine::new();
        let mut context = HashMap::new();
        context.insert("title".to_string(), "Test ADR".to_string());
        context.insert("status".to_string(), "Proposed".to_string());

        let result = engine
            .render("nygard", &context)
            .expect("Template should render successfully");
        assert!(result.contains("# Test ADR"));
        assert!(result.contains("## Status"));
        assert!(result.contains("Proposed"));
    }

    #[test]
    fn test_render_madr() {
        let engine = TemplateEngine::new();
        let mut context = HashMap::new();
        context.insert("title".to_string(), "MADR Test".to_string());
        context.insert("status".to_string(), "accepted".to_string());

        let result = engine.render("madr", &context).expect("Should render");
        assert!(result.contains("# MADR Test"));
        assert!(result.contains("Context and Problem Statement"));
    }

    #[test]
    fn test_render_y_statement() {
        let engine = TemplateEngine::new();
        let mut context = HashMap::new();
        context.insert("title".to_string(), "Y Statement Test".to_string());
        context.insert("status".to_string(), "proposed".to_string());

        let result = engine
            .render("y-statement", &context)
            .expect("Should render");
        assert!(result.contains("# Y Statement Test"));
        assert!(result.contains("In the context of"));
    }

    #[test]
    fn test_render_alexandrian() {
        let engine = TemplateEngine::new();
        let mut context = HashMap::new();
        context.insert("title".to_string(), "Alexandrian Test".to_string());
        context.insert("status".to_string(), "proposed".to_string());

        let result = engine
            .render("alexandrian", &context)
            .expect("Should render");
        assert!(result.contains("# Alexandrian Test"));
        assert!(result.contains("Prologue"));
        assert!(result.contains("Forces"));
    }

    #[test]
    fn test_render_business_case() {
        let engine = TemplateEngine::new();
        let mut context = HashMap::new();
        context.insert("title".to_string(), "Business Case Test".to_string());
        context.insert("status".to_string(), "proposed".to_string());

        let result = engine
            .render("business-case", &context)
            .expect("Should render");
        assert!(result.contains("# Business Case Test"));
        assert!(result.contains("Executive Summary"));
        assert!(result.contains("Cost-Benefit Analysis"));
    }

    #[test]
    fn test_list_templates() {
        let engine = TemplateEngine::new();
        let templates = engine.list_templates();
        assert!(templates.contains(&"nygard".to_string()));
        assert!(templates.contains(&"madr".to_string()));
        assert!(templates.contains(&"y-statement".to_string()));
        assert!(templates.contains(&"alexandrian".to_string()));
        assert!(templates.contains(&"business-case".to_string()));
    }

    #[test]
    fn test_add_custom_template() {
        let mut engine = TemplateEngine::new();
        let custom = "# {{ title }}\n\nCustom template content";
        engine
            .add_template("custom", custom)
            .expect("Should add template");
        assert!(engine.has_template("custom"));
    }

    #[test]
    fn test_get_template() {
        let engine = TemplateEngine::new();
        let nygard = engine.get_template("nygard").expect("Should get template");
        assert!(nygard.contains("## Context"));
    }

    #[test]
    fn test_get_template_not_found() {
        let engine = TemplateEngine::new();
        let result = engine.get_template("nonexistent");
        assert!(result.is_err());
    }

    #[test]
    fn test_render_nonexistent_template() {
        let engine = TemplateEngine::new();
        let context = HashMap::new();
        let result = engine.render("nonexistent", &context);
        assert!(result.is_err());
    }

    #[test]
    fn test_add_template_invalid() {
        let mut engine = TemplateEngine::new();
        // Invalid Tera template syntax
        let invalid = "{% for item in items %}{{ item }}{% endwith %}";
        let result = engine.add_template("invalid", invalid);
        assert!(result.is_err());
    }
}
